crypto/dh/dh_check.c

```diff
@@ -257,22 +257,40 @@ int ossl_dh_check_priv_key(const DH *dh, const BIGNUM *priv_key, int *ret)
     two_powN = BN_new();
     if (two_powN == NULL)
         return 0;
-    if (dh->params.q == NULL)
-        goto err;
-    upper = dh->params.q;
+
+    if (dh->params.q != NULL) {
+        upper = dh->params.q;
+#ifndef FIPS_MODULE
+    } else if (dh->params.p != NULL) {
```

Comment:
This test requires either (dh->params.p != NULL || dh->length != 0)

Reply:
A DH key without p can never be valid.